lib/parallel.rb

```diff
@@ -271,14 +297,22 @@ def map(source, options = {}, &block)
       add_progress_bar!(job_factory, options)
 
       results = if size == 0
-        work_direct(job_factory, options, &block)
-      elsif method == :in_threads
-        work_in_threads(job_factory, options.merge(:count => size), &block)
-      else
-        work_in_processes(job_factory, options.merge(:count => size), &block)
-      end
+                  work_direct(job_factory, options, &block)
+                elsif method == :in_threads
+                  work_in_threads(job_factory, options.merge(:count => size), &block)
+                else
+                  work_in_processes(job_factory, options.merge(:count => size), &block)
+                end
       if results
-        options[:return_results] ? results : source
+        if options[:return_one] 
+          if results.is_a?(Return)
+            results.return_value
+          end
+        elsif options[:return_results] 
+          results
+        else
+          source
+        end
       end
```

Comment:
this looks like it's doing similar kind of logic to `handle_exception` ... can it be combined ?